# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/vm/MemoryMetrics.cpp
# Commit: 480a0b10c49a
# Full Hash: 480a0b10c49a2d97d4332e843ccd62ba2a86449e
# Author: Jon Coppeard <jcoppeard@mozilla.com>
# Date: 2020-10-21 16:00:28
# Regressor Bug: 1657025
# File Overlap Count: 1
# Description:
#   Bug 1657025 - Wait for off-thread parsing to finish before collecting runtime stats r=sfink
#   
#   Another thing that could be causing this mismatch is off-thread parsing, which
#   currently allocates GC things.
#   
# ==============================================================================

diff -r 2ec9908accba -r 480a0b10c49a js/src/vm/MemoryMetrics.cpp
--- a/js/src/vm/MemoryMetrics.cpp	Wed Oct 21 08:10:38 2020 +0000
+++ b/js/src/vm/MemoryMetrics.cpp	Wed Oct 21 08:12:32 2020 +0000
@@ -622,13 +622,17 @@
 static bool CollectRuntimeStatsHelper(JSContext* cx, RuntimeStats* rtStats,
                                       ObjectPrivateVisitor* opv, bool anonymize,
                                       IterateCellCallback statsCellCallback) {
+  // Wait for any off-thread parsing to finish, as that currently allocates GC
+  // things.
+  JSRuntime* rt = cx->runtime();
+  CancelOffThreadParses(rt);
+
   // Finish any ongoing incremental GC that may change the data we're gathering
   // and ensure that we don't do anything that could start another one.
   gc::FinishGC(cx);
   gc::WaitForBackgroundTasks(cx);
   JS::AutoAssertNoGC nogc(cx);
 
-  JSRuntime* rt = cx->runtime();
   if (!rtStats->realmStatsVector.reserve(rt->numRealms)) {
     return false;
   }
